src/CudaSquareImage.cu

```diff
@@ -0,0 +1,58 @@
+/*=========================================================================
+ *
+ *  Copyright NumFOCUS
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *=========================================================================*/
+#include "CudaSquareImage.hcu"
+
+template<class PixelType>
+__global__ void SquareImage3D_kernel(int3 imSize, PixelType* in, PixelType* out)
+{
+  unsigned int i = blockIdx.x * blockDim.x + threadIdx.x;
+  unsigned int j = blockIdx.y * blockDim.y + threadIdx.y;
+  unsigned int k = blockIdx.z * blockDim.z + threadIdx.z;
+
+  if (i >= imSize.x || j >= imSize.y || k >= imSize.z)
+  {
+    return;
+  }
+  unsigned int gidx = i + j * imSize.x + k * imSize.x * imSize.y;
+
+  out[gidx] = in[gidx] * in[gidx];
+}
+
+template<class PixelType>
+void SquareImage3D(int imSize[3], PixelType* in, PixelType* out)
+{
+  // Thread Block Dimensions
+  constexpr int tBlock_x = 16;
+  constexpr int tBlock_y = 4;
+  constexpr int tBlock_z = 4;
+
+  unsigned int blocksInX = (imSize[0] - 1) / tBlock_x + 1;
+  unsigned int blocksInY = (imSize[1] - 1) / tBlock_y + 1;
+  unsigned int blocksInZ = (imSize[2] - 1) / tBlock_z + 1;
+
+  // Compute block and grid sizes
+  dim3 dimGrid = dim3(blocksInX, blocksInY, blocksInZ);
+  dim3 dimBlock = dim3(tBlock_x, tBlock_y, tBlock_z);
+
+  int3 imageSize = make_int3(imSize[0], imSize[1], imSize[2]);
+
+  SquareImage3D_kernel<PixelType><<<dimGrid, dimBlock>>>(imageSize, in, out);
+}
+
+template void SquareImage3D<float>(int imSize[3], float* in, float* out);
+template void SquareImage3D<double>(int imSize[3], double* in, double* out);
```

Comment:
Should these lines be generates by a macro with a list of types configurable by cmake? And is it possible to add a default for all other types?

Reply:
That would be great, but not sure how to define a generic macro that works for all signatures (arguments will be different from one filter to another).

By default you mean using default template? For instance using TImage = itk::Image<float, 3> in itkCudaSquareImageFilter.h?

Reply:
No, sorry, I wasn't clear. I was wondering what happens if you instantiate e.g. `itk::CudaSquareImageFilter< itk::Image<int,2> >`. It actually fails at linking. Another option would be to move the templated code from cu to hcu and to have the compilation fails due to cuda code?

Reply:
I moved it to the hcu file because it feels cleaner. But the error is still fired at linking.

Reply:
I actually meant to move the template definition to the hcu, which you actually need to do since I now get the following error
```
/home/srit/src/ITKCudaCommon/ITKCudaCommon/include/itkCudaSquareImage.hcu: In instantiation of ‘void itk::CudaSquareImage3D(int*, PixelType*, PixelType*) [with PixelType = double]’:
/home/srit/src/ITKCudaCommon/ITKCudaCommon/include/itkCudaSquareImage.hcu:28:79:   required from here
/home/srit/src/ITKCudaCommon/ITKCudaCommon/include/itkCudaSquareImage.hcu:28:79: error: explicit instantiation of ‘void itk::CudaSquareImage3D(int*, PixelType*, PixelType*) [with PixelType = double]’ but no definition available [-fpermissive]
 template void CudaSquareImage3D<double>(int imSize[3], double* in, double* out);
                                                                               ^
```

Reply:
Sorry for the inconvenience, MSVC was dealing fine with this.

But if I move the template code from cu to hcu, the kernel will be used in hcu and the test executable will require nvcc to compile. I don't  think it's what we want.

Reply:
No. Let's keep your first solution, everything in the .cu file.